src/TorchSharp/NN/Module.cs

```diff
@@ -325,55 +208,62 @@ protected internal virtual Module _to(ScalarType dtype)
 
                 protected void _toEpilog(ScalarType dtype)
                 {
-                    foreach (var (_, sm) in named_children()) sm._to(dtype);
+                    _toEpilog(dtype, null);
+                }
 
-                    var alreadyHandled = new HashSet<IntPtr>();
+                protected void _toEpilog(Device device, ScalarType dtype)
+                {
+                    _toEpilog(dtype, device);
+                }
 
-                    foreach (var field in GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)) {
+                protected void _toEpilog(DeviceType deviceType, int deviceIndex)
+                {
+                    _toEpilog(null, new Device(deviceType, deviceIndex));
+                }
 
-                        var fieldName = field.ComponentName();
-                        var value = field.GetValue(this);
+                private void _toEpilog(ScalarType? dtype, Device device)
+                {
+                    foreach (var (_, sm) in named_children()) {
+                        if (device is null) sm._to(dtype.Value);
+                        else if (dtype is null) sm._to(device.type, device.index);
+                        else sm._to(device, dtype.Value);
+                    }
 
-                        switch (value) {
-                        // This order in which these cases are arranged is significant.
-                        case Parameter param when dtype == param.dtype:
-                            alreadyHandled.Add(param.handle);
-                            continue;
-
-                        case Parameter param: {
-                                var t = param.to(dtype);
-                                t.retain_grad();
-                                var p = new Parameter(t, param.requires_grad);
-                                field.SetValue(this, p);
-                                ConditionallyRegisterParameter(fieldName, p);
-                                alreadyHandled.Add(p.handle);
-                                break;
-                            }
+                    var fieldsByComponentName = GetType().GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance)
```

Comment:
Does this have any performance implications?

Reply:
Under the assumption that the costly operation is the reflection, then there shouldn't be. 
We're creating a hash collection of the same size (alreadyHandled vs the Dictionary), so I guess the cost would be a the lookups in the dictionary, which should be very minimal.

Reply:
The brings up a functional concern -- the alreadyHandled set is there to make sure that we don't accidentally deal with the same tensor twice. Is that no longer a possibility?

Reply:
I don't think so. 
The reason it was needed before was because the function was iterating through two different lists of parameters. One using reflection (the parameters registered through the `RegisterComponents()` function), and the other is using the internal list of registered parameters. 
In the proposed code we only go through the list of registered parameters, so there isn't a concern of dealing with the same tensor twice.
Unless there is a case where someone can register the same tensor twice?

Reply:
If someone registers the same parameter under two different names, then we have an issue - the first encounter will dispose the parameter, and then the second time it will have a null tensor error. 
Should this be a use case we should handle?